text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,368 @@
+}
+```
+
+By using `..`, we could ignore the fact that `b` was not visible and move out `a` anyway. This is undesirable, as it would implicitly run `drop(b)` even though `b` was not accessible here. In fact, if there was a requirement that `a` was dropped before `b`, it would never be sound to destructure a type `Bar` in a location where `b` is not accessible.
+
+Finally, you might ask why we need a macro for this at all. This is because for some types, the behavior is slightly different. 
+Let's use the following example. In Rust right now, this does not compile:
+
+```rust
+struct Foo {
+    a: Vec<u64>
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // error: cannot move out of type `Foo`, which implements the `Drop` trait
+    let Foo { a } = x;
+}
+```
+
+This is because we move the field `a` out, and `Foo` implements `Drop`.
+The following does compile:
+
+```rust
+struct Foo {
+    a: Vec<u64>
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // fine, because we don't actually move any fields out
+    let Foo { .. } = x;
+}
+```
+
+Similarly, the following also works because we don't have to move out any fields, we can just copy `a`, because `u64` implements `Copy`.
+
+```rust
+struct Foo {
+    a: u64
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // Totally fine, we just copy a
+    let Foo { a } = x;
+}
+```
+
+In the two examples above, `Drop` still runs for `x`.
+`destructure!()` represents a different operation. When we use it, we actually move the fields out of the type, and prevent `Drop` from running.
+
+```rust
+struct Foo {
+    a: u64
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // `a` is moved out of `x`, Drop never runs for `x`
+    destructure!(let Foo { a } = x);
+}
+```
+
+Preventing `Drop` from running is not inherently problematic in Rust. You can already do this for any type through `mem::forget` (and other means), which is safe.
+
+
+Finally, the following is an example of how one could implement the example from [Motivation](#motivation):
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    let mut this = ManuallyDrop::new(self);
+    let buf = mem::take(&mut this.buf);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    // SAFETY: double-drops are prevented by putting `this` in a ManuallyDrop that is never dropped
+    let inner = unsafe { ptr::read(&this.inner) };
+
+    (inner, buf)
+}
+```
+
+Now with `destructure!`:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    destructure!(let Self { buf, inner, panicked } = self);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    (inner, buf)
+}
+```
+
+We don't even need unsafe anymore :tada:
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
```

Comment:
Something important this is missing: this is adding a new capability to safe code. Unsafe code writers could until today rely on the fact that safe code cannot take ownership of the value of fields in a struct with a `Drop` impl. As such, this RFC is technically a breaking change with soundness implications.

It's probably a rare case that a struct would have important drop-related invariants and be publicly constructible, so this may be acceptable, but this should be noted nonetheless.

Reply:
I believe reliance on `&mut _` not permitting ownership stealing is unsound in presence of `take_mut`, which has proven to be far too useful and natural to disallow.

Reply:
Yea, I'm not sure what this allows in addition to what `take_mut` already allows (which I guess technically isn't blessed, but I would not assume it's cursed either).

In addition to the fact that if _all_ your fields are public, doing dangerous `unsafe` in `Drop` sounds extremely sketchy. In basically all scenarios `Drop` types have all fields private.

Reply:
You can only `take_mut()` if you’re able to produce a replacement value. If there’s no way to get a replacement value for the fields, I think you can’t “steal” a field using `take_mut()`. E.g.

```rust
// crate foo {
pub struct Singleton(());

impl Singleton {
    pub fn claim() -> Self {
        static ONCE: AtomicBool = AtomicBool::new(true);
        if ONCE.swap(false, Relaxed) { Self(()) }
        else { panic!() }
    }
}

pub struct Complex {
    pub field: Singleton,
}

impl Complex {
    pub fn claim() -> Self {
        Self { field: Singleton::claim() }
    }
}

impl Drop for Complex { ... }

// crate bar
let mut complex = Complex::claim();
// not possible with `take_mut`:
// no way to provide replacement value
destruct!(Complex { field } = complex);
```

I’m not sure how important that is though.

Reply:
You can though!
```rust
take_mut(&mut complex.field, |f: Singleton | {
    // temporary ownership
    identity(f)
});

take_mut(&mut complex.field, |f: Singleton | {
    drop(f);
    panic!() // terminate the program
});
```

So at most this breaks something like 
> if a `Drop` type `T` with all public fields is passed to a closure, one of the following happens:
> - `T::drop` is called and all its fields are dropped (modulo panic in drop)
> - `T` is forgotten and none of its fields are dropped
> - `T`'s fields are dropped (/owned) and the program terminates before reaching the end of the closure

And while I assume it's possible to craft a program that requires exactly this assumption, I don't think it's reasonable.

Reply:
> Yea, I'm not sure what this allows in addition to what take_mut already allows

Dropping the fields in a different order I think, though maybe you can do that with nested `take_mut`s?

Reply:
Nested `take_mut` do allow you to drop fields in whichever order you'd like.